pkg/applicationserver/applicationserver.go

```diff
@@ -695,27 +895,29 @@ func (as *ApplicationServer) handleDownlinkQueueInvalidated(ctx context.Context,
 }
 
 func (as *ApplicationServer) handleDownlinkNack(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, msg *ttnpb.ApplicationDownlink, link *link) error {
+	logger := log.FromContext(ctx)
 	client := ttnpb.NewAsNsClient(link.conn)
 	res, err := client.DownlinkQueueList(ctx, &ids, link.callOpts...)
 	if err != nil {
-		log.WithError(err).Warn("Failed to list the downlink queue for inserting nacked downlink message")
+		logger.WithError(err).Warn("Failed to list the downlink queue for inserting nacked downlink message")
 		registerDropDownlink(ctx, ids, msg, err)
 	} else {
-		queue := append([]*ttnpb.ApplicationDownlink{msg}, res.Downlinks...)
 		_, err := as.deviceRegistry.Set(ctx, ids,
 			[]string{
+				"pending_session",
 				"session",
 				"skip_payload_crypto",
 			},
 			func(dev *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
-				if err := as.recalculateDownlinkQueue(ctx, dev, nil, queue, msg.FCnt+1, link); err != nil {
+				queue := append([]*ttnpb.ApplicationDownlink{msg}, res.Downlinks...)
+				if err := as.replaceDownlinkQueue(ctx, dev, link, dev.Session, queue, msg.FCnt+1, true); err != nil {
```

Comment:
I'd expect to see
```go
queue, err := as.recalculateDownlinkQueue(ctx, dev.Session, msg.FCnt+1, queue...)
```
With a following `client.DownlinkQueueReplace` using the returned queue.
Let's allow AS to specify which queue to replace https://github.com/TheThingsNetwork/lorawan-stack/pull/2163#issuecomment-600103266

Reply:
All of the queue operations in the AS have a rollback mechanism - if a queue `replace` fails, the AS attempts to replace the queue in the NS with an empty queue. This rollback boilerplate is part of the `replaceDownlinkQueue`.

The naming itself is not descriptive, and I'll update it.

Reply:
So if the replace fails, AS tries to delete the queue?

Reply:
Yes. This behavior has been here for a long time:
https://github.com/TheThingsNetwork/lorawan-stack/blob/8d8634e3e3f34e466b53eb3f9dcd9af85e96456a/pkg/applicationserver/applicationserver.go#L779-L780
https://github.com/TheThingsNetwork/lorawan-stack/blob/8d8634e3e3f34e466b53eb3f9dcd9af85e96456a/pkg/applicationserver/applicationserver.go#L639-L650